substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,581 @@
+        }
+    }
+
+    protected InvocationResult get(ResolvedJavaMethod caller, Integer bci, ResolvedJavaMethod callee) {
+        return callSitemap(caller).get(Pair.create(bci, toAnalysisMethod(callee)));
+    }
+
+    @Override
+    public void notifyBeforeInline(GraphBuilderContext b, ResolvedJavaMethod methodToInline) {
+        SharedBytecodeParser parser = (SharedBytecodeParser) b;
+        InvocationResultInline inlineData = parser.inlineDuringParsingState;
+        if (inlineData != null) {
+            VMError.guarantee(inlineData.callee == toAnalysisMethod(methodToInline));
+        }
+    }
+
+    @Override
+    public void notifyAfterInline(GraphBuilderContext b, ResolvedJavaMethod methodToInline) {
+        SharedBytecodeParser parser = (SharedBytecodeParser) b;
+        InvocationResultInline inlineData = parser.inlineDuringParsingState;
+        if (inlineData != null) {
+            VMError.guarantee(inlineData.callee == toAnalysisMethod(methodToInline));
+            parser.inlineDuringParsingState = null;
+        }
+    }
+
+    public static AnalysisMethod toAnalysisMethod(ResolvedJavaMethod method) {
+        if (method instanceof AnalysisMethod) {
+            return (AnalysisMethod) method;
+        } else {
+            return ((HostedMethod) method).getWrapped();
+        }
+    }
+
+    /**
+     * Stores information about caller method. This information are used in
+     * {@link InvocationResultInline}.
+     */
+    static final class CallSite {
+        final AnalysisMethod caller;
+        final int bci;
+        final AnalysisMethod callee;
+
+        CallSite(AnalysisMethod caller, int bci, AnalysisMethod callee) {
+            this.caller = caller;
+            this.bci = bci;
+            this.callee = callee;
+        }
+
+        @Override
+        public int hashCode() {
+            return caller.hashCode() * 31 + bci;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            } else if (obj == null || getClass() != obj.getClass()) {
+                return false;
+            }
+            CallSite other = (CallSite) obj;
+            return bci == other.bci && caller.equals(other.caller);
+        }
+
+        @Override
+        public String toString() {
+            return caller.format("%h.%n(%p)") + "@" + bci;
+        }
+    }
+
+    static class InvocationResult {
+        static final InvocationResult ANALYSIS_TOO_COMPLICATED = new InvocationResult();
+
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof InvocationResult;
+        }
+
+        @Override
+        public int hashCode() {
+            return super.hashCode();
+        }
+    }
+
+    public static class InvocationResultInline extends InvocationResult {
+        final CallSite site;
+        final AnalysisMethod callee;
+        final Map<CallSite, InvocationResultInline> children;
```

Comment:
Is the `children` field used? I don't see where the map is populated.

Reply:
The use of this field has been added in the latest version.